LSM: Block Address Determinism

sentientwaffle opened the pull request:
## Background

To enable efficient recovery and additional verification, TigerBeetle's LSM storage will be deterministic.

There are presently three bugs preventing LSM from being deterministic at checkpoint:

1. All concurrent compactions of a half-measure (from different levels & trees) acquire blocks from the FreeSet concurrently.
2. `SuperBlock.Manifest.append()` is called after `ManifestLog` blocks are written. Manifest block writes from different trees are concurrent with one other, so this is nondeterministic.
3. `ManifestLog.append()` is called concurrently by all running level compactions within tree.

This PR fixes **1** and **2**.
**3** is deferred for a future PR (see below).

## 1: Block Reservation

Previously all concurrent compactions acquire block addresses on an as-needed basis (for both output tables and manifest log blocks).

This is convenient, but because compaction progress is nondeterministic (based on IO ordering) block acquisition order — and therefore storage in general — is likewise nondeterministic.

To fix this:
- Each compaction (per tree/level) reserves an upper bound of block addresses that it may require for compaction.
- The manifest log also reserves the upper bound of block addresses that it may require for compaction.

A compaction or manifest log then acquires a subset of the blocks from its reservation. These acquisitions may be concurrent _between reservations_ without compromising storage determinism. (Within a reservation, acquisitions must still be sequential).

The "lifecycle" of a reservation is documented in `superblock_free_set.zig`.

## 2: `SuperBlock.Manifest.append()`

`SuperBlock.Manifest.append()` is now handled by `ManifestLog.flush()`, which in turn is called synchronously relative to the start of a compaction-beat or checkpoint.

## 3: `ManifestLog.append()`

The `ManifestLog.append()` nondeterminism is _not resolved by this PR._ It will be fixed for free by compaction pacing, assuming we compact the levels of each tree sequentially (with trees still concurrent). (Sequential compaction of levels is also friendlier to SSD garbage collection).

Because this issue isn't fixed yet, the VOPR's `StorageChecker` remains disabled.

## Testing

- `FreeSet` fuzzer:
  - Fuzz the reserve/acquire/forfeit/release lifecycle.
  - Encoding/decoding is covered by the EWAH fuzzer (and extended in an upcoming PR).
- `ManifestLog` fuzzer:
  - Extended to verify `SuperBlock.Manifest.compaction_set` membership.
  - Don't interleave more `append()`s between `compact()`s than the actual Tree will need.
- `Tree`/`Forest` fuzzers: See `8749454`.

## Future Work

In addition to fixing `ManifestLog.append()` determinism, per-beat compaction determinism will enable stricter (smaller) block reservations, which will reduce block fragmentation.

sentientwaffle committed:
SuperBlock: FreeSet reserve/acquire_reserved

Currently all concurrent compactions acquire block addresses on an as-needed basis (for both output tables and manifest log blocks).

This is convenient, but because compaction progress is nondeterministic (based on IO ordering) the block acquisition (from each tree's perspective) — and therefore storage in general — is likewise nondeterministic.

To fix this, my plan is that each compaction will (serially) reserve an upper bound of block addresses that it may require for compaction. Then during compaction it acquires address from the reservation — since that is per-tree, the order is deterministic. After the compaction measure (or beat, once beat compaction pacing is implemented), it will release any leftover reserved addresses.

This also ensures that there are no outstanding reservations at checkpoint, so we will recover to the pre-crash state (i.e. choose the same reservations).

sentientwaffle committed:
SuperBlock: Fuzz FreeSet

Only fuzz the reservation API, since I am removing the old API shortly.

sentientwaffle committed:
LSM: Document lsm_table_size_max.

sentientwaffle committed:
StateMachine: Fix batch_max invariant typo

sentientwaffle committed:
SuperBlock: FreeSet: Change API to reserve/acquire

Remove the old API acquire/release API, switching to the new "reservation" API.
This isn't fully integrated yet, i.e. the concurrent block allocation issue isn't fixed yet.

Each `FreeSet.reserve()` now corresponds to a `FreeSet.forfeit()`. The original implementation used a single `forfeit()` to free all reservations. For efficient implementation we still don't want to allow mixed reserve/forfeit workloads, but making them 1:1 means that each test/fuzzer doesn't need to independently manage the freeset, which would have created a strange dependency & effectively increased the API surface area of Forest/Groove/Tree/ManifestLog/Compaction.

- Rename "AddressReservation" to "Reservation"; it is unambiguous within the freeset (or grid) namespace.
- `FreeSet.count_free_reserved`: This will be used for assertions, to check that reservations have enough blocks remaining at various points throughout compaction.
- `FreeSet.count_reservations()`: Used for assertions, so that the replica can verify that all reservations are freed.

Also, change `count_free()` and `count_acquired()` to return a `usize` instead of `u64`. Reasoning:
- This matches the underlying `BitSet`'s API. This is good from a consistency perspective, but also to avoid giving the false impression that the bitset can handle more than `max(u32)` blocks on a 32-bit machine.
- Avoids possible confusion with the `u64` address type.

sentientwaffle committed:
LSM/ManifestLog: Compute buffered block count

Instead of hard-coding the buffered block count, compute it using the worst-case of compaction.
The result is the same for our current config, but it will now work properly with a smaller block size.

Also:
- Fix fuzzer: Previously during event generation, the `entry_count` was being updated incorrectly. It worked anyway because there were two mistakes that cancelled each other out!
- Improve fuzzer: generate fewer `noop` events.

sentientwaffle committed:
SuperBlock/FreeSet: Track reservation session

This allows the FreeSet to verify that reservations are forfeited and allocated from during the correct period.

sentientwaffle committed:
LSM: Don't compact the first measure

Skip compaction for the first measure (`op < lsm_batch_multiple`).
This previously worked (accidentally, I think) because all of the compactions were idle. But it creates a strange asymmetry for the first half-bar because first beat never occurs, but the last one does. This will cause problems once the ManifestLog reserves blocks at the first beat and removes them at the last.

sentientwaffle committed:
LSM: Fix small tables

Smaller tables require:
- More iterations when solving the comptime constraint problem.
- `block_key_count` was -1 when it should have been 0.
- `block_key_count == 0` was not handled correctly. (The compiler complained about trying to `@alignCast` an empty slice.
- A single filter block may cover more data blocks than the table contains.

sentientwaffle committed:
LSM: Reserve blocks before acquire

Compactions of multiple levels and trees run concurrently. To ensure storage is deterministic, all blocks written by those compactions (index/filter/data blocks from output tables, and manifest log blocks) must likewise be assigned in a deterministic order.

To achieve this, at the beginning of each half-bar, all compactions & manifest logs synchronously reserve all the blocks they will need (in the worst case). Then each concurrent compaction (or manifest log) acquires blocks *only from their own reservation*.

To visualize, consider the following operations:

  1. Compaction A acquires a block.
  2. Compaction B acquires a block.
  3. Compaction A acquires a block.
  4. Compaction B acquires a block.

This could result in many different block assignments (depending on IO ordering by the OS):

    1 2 3 4  (address)
    A B A B
    B A B A
    ...etc

With address reservations, it is guaranteed to be:

    1 2 3 4  (address)
    A B A B

Block releases are not impacted by this issue; they are only made available for re-acquisition at checkpoint (see: `FreeSet.staging`).

This commit should fix storage determinism at time-of-checkpoint and compaction, but the storage checker isn't re-enabled yet — the manifest log acquires blocks before it is ready to write them (it acquires at time of `close_block()`), which throws off the on-compact storage checker.

sentientwaffle committed:
LSM/ManifestLog: Fix storage determinism issue

Problem: After a manifest log block was written, its tree/address/checksum is appended to the superblock manifest list. Concurrent manifest log flushes race each other for positions in the superblock manifest.

Solution: Append the information of each closed to the superblock manifest before they are written. These updates are synchronous relative to one another (i.e. relative to `ManifestLog.compact()`'s or `ManifestLog.checkpoint()`'s invokation).

This bug was caught by the StorageChecker. With this fix, storage is now deterministic at checkpoint, but not yet at the compaction-bar boundary. (Or rather, it is deterministic, but the manifest log holds acquired block addresses before they are written to, so the StorageChecker can't use the free set to identify active blocks for verification).

sentientwaffle committed:
VOPR/StorageChecker: Don't check WAL headers

The WAL headers are not deterministic because the Journal writes an invalid (checksum=0 + reserved) header to the redundant headers when a `faulty` bit is set. And the repair may not have occurred by the time of checkpoint if the entry was marked faulty _after_ it was already committed.

sentientwaffle committed:
LSM/ManifestLog/Fuzz: Fix appends-per-compact

Change the ManifestLog fuzzer to better approximate actual usage w.r.t. the number of appends (insert/remove) between each `compact()`.
This was wrong before — it would append too many times, and then there wasn't a block left in the reservation for the ManifestLog compaction.

sentientwaffle committed:
LSM/ManifestLog/Fuzz: Verify compaction queue

After each checkpoint, verify that the compaction queue (`SuperBlock.Manifest.compaction_set`) contains exactly what it should.

The ManifestLog changes are mostly to export the Block helpers.

wip

sentientwaffle committed:
LSM/Tree: Compaction overlaps on zero/one edge

Additionally, verify that a compaction never outputs more tables than it had as input.

sentientwaffle committed:
LSM/ManifestLog: Tighten reservation upper bound

Tighten the ManifestLog's FreeSet block-count reservation. It only needs to reserve 2 blocks (even though it has a maximum of 3 buffered) because to hit 3 buffered, it must already have 1 closed, so it does not need to reserve an extra address.

Tightening the upper bound as much as possible minimizes space amplification due to freeset fragmentation.

sentientwaffle committed:
LSM/VSR: Clarify code comments

jamii commented on src/vsr/superblock_free_set.zig:
```diff
@@ -96,18 +159,102 @@ pub const FreeSet = struct {
+    }
+
+    /// Marks a free block from the reservation as allocated, and returns the address.
+    /// The reservation must not have been forfeited — that is, they must have been no calls to
```
This sentence is incomplete.

jamii replied:
Also probably s/they/there

jamii commented on src/vsr/superblock_free_set.zig:
```diff
@@ -96,18 +159,102 @@ pub const FreeSet = struct {
+            set.index.bit_length,
+        ) orelse return null;
+
+        var block = std.math.max(shard_start * shard_size, set.reservation_blocks);
```
It wasn't obvious to me at first that Reservation is not a contiguous range but instead promises that there are enough free blocks in that range. That made this code hard to follow. Maybe a comment along the lines of `Find a range which contains reserve_count free blocks`.

sentientwaffle replied:
:+1: I'll make that more clear. There is a note [here](https://github.com/tigerbeetledb/tigerbeetle/blob/5c123869540ad5c45c88171b869cb0a932f20ff3/src/vsr/superblock_free_set.zig#L166-L167) too:

```zig
///   - If a reservation is returned, it covers exactly `reserve_count` free blocks, along with
///     any interleaved already-acquired blocks.
```

jamii commented:
I cranked up the memory for the forest fuzzer. 

```
> git diff
diff --git a/src/lsm/forest_fuzz.zig b/src/lsm/forest_fuzz.zig
index 39590554..7479075e 100644
--- a/src/lsm/forest_fuzz.zig
+++ b/src/lsm/forest_fuzz.zig
@@ -41,7 +41,7 @@ const Environment = struct {
     const size_max = vsr.Zone.superblock.size().? +
         vsr.Zone.wal_headers.size().? +
         vsr.Zone.wal_prepares.size().? +
-        1024 * 1024 * 1024;
+        16 * 1024 * 1024 * 1024;
```

This seed fails on your branch but not on main:

```
> zig build fuzz_lsm_forest -Drelease-safe=true -- --seed 14709536695519122948
info(fuzz): Fuzz seed = 14709536695519122948
info(lsm_forest_fuzz): fuzz_op_count = 43641
info(lsm_forest_fuzz): fuzz_op_distribution = std.enums.struct:24:19{ .compact = 1.00e+00, .put_account = 8.00e+00, .get_account = 0.00e+00 }
info(lsm_forest_fuzz): puts_since_compact_max = 8191
info(lsm_forest_fuzz): compacts_per_checkpoint = 256
thread 19935 panic: reached unreachable code
/home/jamie/tigerbeetle/src/lsm/groove.zig:0:0: 0x250541 in lsm.groove.GrooveType(test.storage.Storage,tigerbeetle.Account,(struct state_machine.struct:35:14 constant)).compact (fuzz_lsm_forest)
/home/jamie/tigerbeetle/src/lsm/forest.zig:191:59: 0x22872d in Environment.run (fuzz_lsm_forest)
                @field(forest.grooves, field.name).compact(Join.groove_callback(field.name), op);
                                                          ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:295:24: 0x224690 in main (fuzz_lsm_forest)
    try Environment.run(&storage, fuzz_ops);
                       ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:561:37: 0x210181 in std.start.posixCallMainAndExit (fuzz_lsm_forest)
            const result = root.main() catch |err| {
                                    ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:322:5: 0x20fb8e in std.start._start (fuzz_lsm_forest)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^
The following command terminated unexpectedly:
cd /home/jamie/tigerbeetle && /home/jamie/tigerbeetle/zig-cache/o/5f5d2780ae3cfbc3acbba41e15898995/fuzz_lsm_forest --seed 14709536695519122948
error: the following build command failed with exit code 1:
/home/jamie/tigerbeetle/zig-cache/o/18e602c2a5fae2d80096678abf003c6a/build /nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/bin/zig /home/jamie/tigerbeetle /home/jamie/tigerbeetle/zig-cache /home/jamie/.cache/zig fuzz_lsm_forest -Drelease-safe=true -- --seed 14709536695519122948
```

jamii commented:
It's a confusing crash. It remains if I rebase your branch on master, so it's not something that was recently fixed. But it isn't obviously linked to the reserve changes:

```
thread 20782 panic: reached unreachable code
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/debug.zig:225:14: 0x24250b in std.debug.assert (fuzz_lsm_forest)
    if (!ok) unreachable; // assertion failure
             ^
/home/jamie/tigerbeetle/src/lsm/tree.zig:541:19: 0x2b3a07 in lsm.tree.TreeType(lsm.table.TableType(u128,lsm.groove.IdTreeValue,lsm.groove.IdTreeValue.compare_keys,lsm.groove.IdTreeValue.key_from_value,554112867134706473364364839029663282043,lsm.groove.IdTreeValue.tombstone,lsm.groove.IdTreeValue.tombstone_from_key),test.storage.Storage,[]const u8{65,99,99,111,117,110,116,46,105,100}).compact (fuzz_lsm_forest)
            assert(op == tree.lookup_snapshot_max);
                  ^
/home/jamie/tigerbeetle/src/lsm/groove.zig:946:31: 0x2a3e13 in lsm.groove.GrooveType(test.storage.Storage,tigerbeetle.Account,(struct state_machine.struct:35:14 constant)).compact (fuzz_lsm_forest)
            groove.ids.compact(Join.tree_callback(.ids), op);
                              ^
/home/jamie/tigerbeetle/src/lsm/forest.zig:191:59: 0x295391 in lsm.forest.ForestType(test.storage.Storage,(struct state_machine.struct:50:49 constant)).compact (fuzz_lsm_forest)
                @field(forest.grooves, field.name).compact(Join.groove_callback(field.name), op);
                                                          ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:182:27: 0x294c00 in Environment.compact (fuzz_lsm_forest)
        env.forest.compact(forest_compact_callback, op);
                          ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:260:32: 0x2875a8 in Environment.run (fuzz_lsm_forest)
                    env.compact(compact.op);
                               ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:295:24: 0x274826 in run_fuzz_ops (fuzz_lsm_forest)
    try Environment.run(&storage, fuzz_ops);
                       ^
/home/jamie/tigerbeetle/src/lsm/forest_fuzz.zig:405:21: 0x26c446 in main (fuzz_lsm_forest)
    try run_fuzz_ops(Storage.Options{
                    ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:561:37: 0x26579a in std.start.callMain (fuzz_lsm_forest)
            const result = root.main() catch |err| {
                                    ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:495:12: 0x24638e in std.start.callMainWithArgs (fuzz_lsm_forest)
    return @call(.{ .modifier = .always_inline }, callMain, .{});
           ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:409:17: 0x245426 in std.start.posixCallMainAndExit (fuzz_lsm_forest)
    std.os.exit(@call(.{ .modifier = .always_inline }, callMainWithArgs, .{ argc, argv, envp }));
                ^
/nix/store/0hn8yvvw94r2q1hihdfvy1s9j3sy431g-zig/lib/std/start.zig:322:5: 0x245232 in std.start._start (fuzz_lsm_forest)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^
```

jamii commented on src/lsm/forest_fuzz.zig:
```diff
@@ -199,11 +201,12 @@ const Environment = struct {
         env.change_state(.forest_checkpointing, .superblock_checkpointing);
         env.superblock.checkpoint(superblock_checkpoint_callback, &env.superblock_context, .{
             .commit_min_checksum = env.superblock.working.vsr_state.commit_min_checksum + 1,
-            .commit_min = env.superblock.working.vsr_state.commit_min + 1,
```
Oh, this is suspicious. Maybe the cause of the fuzzer crash?

sentientwaffle replied:
Yep, should be

```zig
.commit_min = env.checkpoint_op.?,
```

instead of

```zig
.commit_min = env.checkpoint_op.? + 1,
```

jamii commented on src/lsm/manifest_log.zig:
```diff
@@ -78,29 +77,44 @@ pub fn ManifestLogType(comptime Storage: type, comptime TableInfo: type) type {
+        /// compaction (not including manifest log compaction).
+        ///
+        /// Input tables are removed from the manifest.
+        /// Output tables are removed from the manifest.
```
Output tables are inserted into the manifest?

sentientwaffle replied:
Thanks, typo.

jamii commented on src/lsm/tree.zig:
```diff
@@ -51,6 +51,21 @@ const half_bar_beat_count = @divExact(config.lsm_batch_multiple, 2);
+/// +1: from level B, with one edge overlapping the lsm_growth_factor tables.
+///     (This is not +2 because input table selection is least-overlap. If the input table overlaps
+///     on both edges, there must be another table with less overlap to select).
+pub const compaction_tables_input_max = 1 + config.lsm_growth_factor + 1;
```
I'm still confused about why this is +1 and not +0. 

Suppose every table in level A overlaps at least growth_factor + 1 tables in level B.
Assign each table in level B to the first table in level A that overlaps it.
Then the first table in level A has at least growth_factor + 1 tables assigned to it, and every other table in level A has at least growth_factor tables assigned.
So the total number of tables in level B must be at least (level_a_table_count * growth_factor) + 1, which is too many.

sentientwaffle replied:
You're correct; I just didn't understand that before :+1:.

jamii commented:
Everything else looks fine.

sentientwaffle committed:
LSM: Fix fuzzer, ManifestLog upper-bound

- A compaction input table from level A overlaps _at most_ `lsm_growth_factor` tables from level B.
- Fix an off-by-one in `forest_fuzz.zig`: https://github.com/tigerbeetledb/tigerbeetle/pull/247#discussion_r1028444657
- In ManifestLog, input blocks are updated, then removed (in the same compaction).
